src/java.base/share/classes/java/lang/reflect/code/bytecode/UnresolvedType.java

```diff
@@ -0,0 +1,202 @@
+/*
+ * Copyright (c) 2024, Oracle and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.  Oracle designates this
+ * particular file as subject to the "Classpath" exception as provided
+ * by Oracle in the LICENSE file that accompanied this code.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+package java.lang.reflect.code.bytecode;
+
+import java.lang.reflect.code.TypeElement;
+import java.lang.reflect.code.type.ArrayType;
+import java.lang.reflect.code.type.JavaType;
+import java.util.List;
+
+sealed interface UnresolvedType extends TypeElement {
+
+    static Ref unresolvedRef() {
+        return new Ref();
+    }
+
+    static Int unresolvedInt() {
+        return new Int();
+    }
+
+    JavaType resolved();
+
+    boolean resolveTo(TypeElement type);
+    boolean resolveFrom(TypeElement type);
+
+    static final class Ref implements UnresolvedType {
+        private static final TypeElement.ExternalizedTypeElement UNRESOLVED_REF = new TypeElement.ExternalizedTypeElement("?REF", List.of());
+
+        private JavaType resolved;
```

Comment:
Sneaky mutability :-) Not something i would recommend although i understand the convenience. You could instead use a map of unresolved type to resolved type. But let's consider that later as i think there is a broader issue. More generally the transformation of types is harder than it should be, and this is a good use case to help improve this area. I don't have any solid ideas right now but hopefully we can eventually come up with something better.

Reply:
Yes, this is not ideal. Maps can be used later in the transform, however unresolved component of an unresolved array type is something that needs to be at least partially resolved (2-slot long or double vs 1-slot others) in the first stage of BytecodeLift or it causes stack problems with hybrid single and dual slot instructions (dup_x2, dup2, dup2_x1 and dup2_x2).
I'll definitelly look for better solution.

Reply:
Maybe I can inject an explicit ConvOp after ArrayLoadOp result of an unresolved type, however it will generate an instability. Or custom ResolveTypeOp, which can be later removed by a transform.

Reply:
Common solution for `componentType` and `ArrayLoadOp` loosing information during lift would be 
`ArrayLoadOp` accepting result component type as an additional optional argument.
In that case we don't have to add `TypeWithComponent` interface and `BytecodeLift` will inject specific (resolved or unresolved) type as a result of `ArrayLoadOp`. It would allow to remove `UnresolvedType.Comp` and the mutable field and move the type resolution completely to the `UnresolvedTypesTransformer`.

Reply:
That sounds like a better approach.

Reply:
OK, first step done. Now I can move the resolution field and logic out.

Reply:
All fields and logic moved from UnresolvedType and SlotOp.